### PR TITLE
ALGO-701Scala IPA language support

### PR DESCRIPTION
**Pull request by zeryx**
Updated the Scala client to contain an Algorithm Handler; this new class operates and manages algorithms in the new IPA algorithm development language support system. 
For more info check out IPA confluence

### CODE REVIEW

File: src/main/scala/com/algorithmia/handler/AbstractAlgorithm.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+   implicit class WeakTypeDetector[I: WeakTypeTag](related: AbstractAlgorithm[I, _]) {
+     def getType: Type = {
+       weakTypeOf[I]
+     }
+   }
+   implicit class TypeDetector[I: TypeTag](related: AbstractAlgorithm[I, _]) {
+     def getType: Type = {
+       typeOf[I]
+     }
+   }
```
Comment (zeryx) [inline src/main/scala/com/algorithmia/handler/AbstractAlgorithm.scala]: These seem to ensure that we're always able to get the input type successfully, including if it's a complex type or primitive type. My hunch is, can we reuse this same style of logic to automagically get the Reads/Writes implicit values if they exist, and if not - generate them? @kennydaniel what do you think?

File: src/main/scala/com/algorithmia/handler/RequestHandler.scala [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+import play.api.libs.json._
+
+import scala.reflect.ClassTag
+import scala.util._
+
+case class RequestHandler[I]() {
+
+  def convertToInput(input: JsValue)(implicit s: Reads[I], obj: ClassTag[I]): Try[I] = {
+    Json.fromJson[I](input).asOpt match {
```
Comment (kennydaniel) [inline src/main/scala/com/algorithmia/handler/RequestHandler.scala]: this method still feels unnecessary... its basically mapping success to success, and failure to failure.

Comment (zeryx) [reply src/main/scala/com/algorithmia/handler/RequestHandler.scala]: it has a custom error message and I felt like the actual operation of converting from a JsValue to an I was significant enough to warrant it's own function. But I could see it living as a lambda in processRequest

**Review by besirkurtulmus — APPROVED** [review]
LGTM.

File: src/main/scala/com/algorithmia/handler/ResponseHandler.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+  private def write(data: String): Try[Unit] = {
+    Try(new PrintStream(new FileOutputStream(this.FIFOPATH, true)))
+      .map({ s =>
+        s.println(data); s
+      })
+      .map({ s =>
+        s.flush(); s
+      })
+      .map(_.close())
+  }
```
Comment (kennydaniel) [inline src/main/scala/com/algorithmia/handler/ResponseHandler.scala:19]: these should not be 3 map statements. It should probably just be one:
```
    Try(new PrintStream(new FileOutputStream(this.FIFOPATH, true)))
      .map(s => {
        s.println(data)
        s.flush
        s.close
    })
```

**Comment by kennydaniel** [comment]
scalafmt needs to be re-run on RequestHandler. 
There's some code i'd like to make more idiomatic scala.

**Comment by kennydaniel** [comment]
Also these compiler warnings seem suspicious:
```
[warn] /home/kenny/workspace/clients/algorithmia-scala/src/main/scala/com/algorithmia/handler/Response.scala:14:31: parameter value writeO in method ww is never used
[warn]   implicit def ww[O](implicit writeO: Writes[O]): Writes[Response[O]] = Json.writes[Response[O]]
[warn]                               ^
[warn] /home/kenny/workspace/clients/algorithmia-scala/src/main/scala/com/algorithmia/handler/ResponseHandler.scala:30:66: parameter value obj in method writeResponseToPipe is never used
[warn]   def writeResponseToPipe(output: O)(implicit writer: Writes[O], obj: ClassTag[O]): Try[Unit] = {
[warn]       
```

**Comment by zeryx** [comment]
> Also these compiler warnings seem suspicious:
> 
> ```
> [warn] /home/kenny/workspace/clients/algorithmia-scala/src/main/scala/com/algorithmia/handler/Response.scala:14:31: parameter value writeO in method ww is never used
> [warn]   implicit def ww[O](implicit writeO: Writes[O]): Writes[Response[O]] = Json.writes[Response[O]]
> [warn]                               ^
> [warn] /home/kenny/workspace/clients/algorithmia-scala/src/main/scala/com/algorithmia/handler/ResponseHandler.scala:30:66: parameter value obj in method writeResponseToPipe is never used
> [warn]   def writeResponseToPipe(output: O)(implicit writer: Writes[O], obj: ClassTag[O]): Try[Unit] = {
> [warn]       
> ```

If I replace the variable with a `_` I get compiler failures around missing implicits, I'm not sure when/where implicits are actually considered "used" in cases like this. Is there a better system to manage that?

File: src/main/scala/com/algorithmia/handler/ResponseHandler.scala (hunk 3 above)
Comment (zeryx) [reply src/main/scala/com/algorithmia/handler/ResponseHandler.scala:19]: accepted, agreed this is cleaner.

**Review by kennydaniel — APPROVED** [review]

**Pull request merged**